cli/internal/fs/copy_file.go

```diff
@@ -28,43 +32,44 @@ func CopyOrLinkFile(from, to string, fromMode, toMode os.FileMode, link, fallbac
 			}
 			return os.Symlink(dest, to)
 		}
-		if err := os.Link(from, to); err == nil || !fallback {
+		if err := os.Link(from.Path.ToString(), to); err == nil || !fallback {
 			return err
 		}
 	}
-	return CopyFile(from, to, toMode)
+	return CopyFile(from, to)
 }
 
 // RecursiveCopy copies either a single file or a directory.
 // 'mode' is the mode of the destination file.
-func RecursiveCopy(from string, to string, mode os.FileMode) error {
-	return RecursiveCopyOrLinkFile(from, to, mode, false, false)
+func RecursiveCopy(from string, to string) error {
+	return RecursiveCopyOrLinkFile(from, to, false, false)
 }
 
 // RecursiveCopyOrLinkFile recursively copies or links a file or directory.
-// 'mode' is the mode of the destination file.
 // If 'link' is true then we'll hardlink files instead of copying them.
 // If 'fallback' is true then we'll fall back to a copy if linking fails.
-func RecursiveCopyOrLinkFile(from string, to string, mode os.FileMode, link, fallback bool) error {
-	info, err := os.Lstat(from)
+func RecursiveCopyOrLinkFile(from string, to string, link bool, fallback bool) error {
+	statedFrom := LstatCachedFile{Path: UnsafeToAbsolutePath(from)}
+	fromInfo, err := statedFrom.GetInfo()
 	if err != nil {
 		return err
 	}
-	if info.IsDir() {
-		return WalkMode(from, func(name string, isDir bool, fileMode os.FileMode) error {
-			dest := filepath.Join(to, name[len(from):])
+
+	if fromInfo.IsDir() {
+		return WalkMode(statedFrom.Path.ToString(), func(name string, isDir bool, fileType os.FileMode) error {
+			dest := filepath.Join(to, name[len(statedFrom.Path.ToString()):])
 			if isDir {
 				return os.MkdirAll(dest, DirPermissions)
 			}
-			if isSame, err := SameFile(from, name); err != nil {
+			if isSame, err := SameFile(statedFrom.Path.ToString(), name); err != nil {
```

Comment:
I don't know how many places call `SameFile`, but if they're all in this file and tests, passing a `LstatCachedFile` in would save a stat.

If they're not all here then don't worry about it.